users/shared_account/invite.go

```diff
@@ -191,46 +166,62 @@ func sendMailNotification(ctx context.Context, tx *sql.Tx, userMail string, user
 	return nil
 }
 
+func inviteUserAlreadyExist(ctx context.Context, tx *sql.Tx, body InviteUserRequest, guestId int) (int, interface{}) {
+	logger := jsonlog.LoggerFromContextOrDefault(ctx)
+	isAlreadyShared, err := checkSharedAccount(ctx, tx, body.AccountId, guestId)
+	if err != nil {
+		return 403, ErrorInviteUser
+	} else if isAlreadyShared {
+		return http.StatusBadRequest, ErrorAlreadyShared
+	}
+	sharedAccount, err := addAccountToGuest(ctx, tx, body.AccountId, body.PermissionLevel, guestId)
+	if err == nil {
+		err = sendMailNotification(ctx, tx, body.Email,true, 0)
+		if err != nil {
+			logger.Error("Error occured while sending an email to an existing user.", err.Error())
+			return 403, ErrorInviteUser
+		}
+		return http.StatusBadRequest, sharedAccount
+	} else {
+		logger.Error("Error occured while adding account to an existing user.", err.Error())
+		return 403, ErrorInviteUser
+	}
+}
+
+func inviteNewUser(ctx context.Context, tx *sql.Tx, body InviteUserRequest) (int, interface{}) {
+	logger := jsonlog.LoggerFromContextOrDefault(ctx)
+	newUserId, newUser, err := createAccountForGuest(ctx, tx, body.Email, body.AccountId, body.PermissionLevel)
+	if err == nil {
+		err = sendMailNotification(ctx, tx, body.Email,false, newUserId)
+		if err != nil {
+			logger.Error("Error occured while sending an email to a new user.", err.Error())
+			return 403, ErrorInviteNewUser
+		}
+		return http.StatusBadRequest, newUser
+	} else {
+		logger.Error("Error occured while creating new account for a guest.", err.Error())
+		return 403, ErrorInviteNewUser
+	}
+}
+
 // inviteUserWithValidBody tries to share an account with a specific user
-func inviteUserWithValidBody(request *http.Request, body inviteUserRequest, tx *sql.Tx, user User) (int, interface{}) {
+func InviteUserWithValidBody(request *http.Request, body InviteUserRequest, tx *sql.Tx, user users.User) (int, interface{}) {
 	logger := jsonlog.LoggerFromContextOrDefault(request.Context())
-	result, guestId, err := checkUserWithEmail(request.Context(), tx, body.Email)
-	if err == nil {
+	security, err := safetyCheckByAccountId(request.Context(), tx, body.AccountId, user)
+	if !security || err != nil {
+		return 403, err
+	}
+	result, guestId, err := checkUserWithEmail(request.Context(), tx, body.Email, user)
+	if err == nil && (body.PermissionLevel == 0 || body.PermissionLevel == 1 || body.PermissionLevel == 2) {
```

Comment:
You should create a function to test the permission level. If you want to change it, you would just have to modify one function. And it is not really clean a such `if`